src/main/groovy/util/BufferedIterator.java

```diff
@@ -0,0 +1,69 @@
+/**
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package groovy.util;
+
+import java.util.Iterator;
+
+/**
+ * An iterator that allows examining the next element without consuming it.
+ *
+ * @author Andrew Taylor
+ * @since 2.5.0
+ */
+public class BufferedIterator<T> implements Iterator<T> {
+
+    private final Iterator<T> iter;
+    private boolean hasBuffered;
+    private T buffered;
+
+    public BufferedIterator(Iterator<T> iter) {
+        this.iter = iter;
+        this.hasBuffered = false;
+    }
+
+    public boolean hasNext() {
+        return hasBuffered || iter.hasNext();
+    }
+
+    public T next() {
+        if (hasBuffered) {
+            T buffered = this.buffered;
+            this.buffered = null;
+            hasBuffered = false;
+            return buffered;
+        } else {
+            return iter.next();
+        }
+    }
+
+    public void remove() {
+        throw new UnsupportedOperationException();
```

Comment:
If the `head` method has been called and an item is buffered, the underlying iterator is pointing at the wrong element.  Delegating would remove the item after the one it should.  This case could be detected but there's no way to handle it correctly.  Rather than failing in some cases, I think it's better to just not support removal.